Sources/GRPC/ConnectionPool/PoolManager.swift

```diff
@@ -0,0 +1,314 @@
+  /// Each connection pool is configured using the `perPoolConfiguration`.
+  ///
+  /// - Parameters:
+  ///   - group: The `EventLoopGroup` providing `EventLoop`s to connections managed by the pool
+  ///       manager.
+  ///   - perPoolConfiguration: Configuration used by each connection pool managed by the manager.
+  ///   - logger: A logger.
+  /// - Returns: An initialized pool manager.
+  internal static func makeInitializedPoolManager(
+    using group: EventLoopGroup,
+    perPoolConfiguration: PerPoolConfiguration,
+    logger: GRPCLogger
+  ) -> PoolManager {
+    let manager = PoolManager(group: group)
+    manager.initialize(perPoolConfiguration: perPoolConfiguration, logger: logger)
+    return manager
+  }
+
+  private init(group: EventLoopGroup) {
+    self.state = PoolManagerStateMachine(.inactive)
+    self.pools = []
+    self.group = group
+
+    // The pool relies on the identity of each `EventLoop` in the `EventLoopGroup` being unique. In
+    // practice this is unlikely to happen unless a custom `EventLoopGroup` is constructed, because
+    // of that we'll only check when running in debug mode.
+    debugOnly {
+      let eventLoopIDs = group.makeIterator().map { ObjectIdentifier($0) }
+      let uniqueEventLoopIDs = Set(eventLoopIDs)
+      assert(
+        eventLoopIDs.count == uniqueEventLoopIDs.count,
+        "'group' contains non-unique event loops"
+      )
+    }
+  }
+
+  deinit {
+    self.lock.withLockVoid {
+      assert(
+        self.state.isShutdownOrShuttingDown,
+        "The pool manager (\(ObjectIdentifier(self))) must be shutdown before going out of scope."
+      )
+    }
+  }
+
+  /// Initialize the pool manager, create and initialize one connection pool per event loop in the
+  /// pools `EventLoopGroup`.
+  ///
+  /// - Important: Must only be called once.
+  /// - Parameters:
+  ///   - configuration: The configuration used for each connection pool.
+  ///   - logger: A logger.
+  private func initialize(
+    perPoolConfiguration configuration: PerPoolConfiguration,
+    logger: GRPCLogger
+  ) {
+    var logger = logger
+    logger[metadataKey: Metadata.id] = "\(ObjectIdentifier(self))"
+
+    assert(self.pools.isEmpty)
+    self.pools = self.makePools(perPoolConfiguration: configuration, logger: logger)
+
+    logger.debug("initializing connection pool manager", metadata: [
+      Metadata.poolCount: "\(self.pools.count)",
+      Metadata.connectionsPerPool: "\(configuration.maxConnections)",
+      Metadata.waitersPerPool: "\(configuration.maxWaiters)",
+    ])
+
+    // The assumed maximum number of streams concurrently available in each pool.
+    let assumedCapacity = configuration.assumedStreamCapacity
+
+    // The state machine stores the per-pool state keyed by the pools `EventLoopID` and tells the
+    // pool manager about which pool to use/operate via the pools index in `self.pools`.
+    let poolKeys = self.pools.enumerated().map { poolIndex, pool in
```

Comment:
A minor defensiveness note: `poolIndex` is an `index` _only_ if the stored data type is `Array`. Otherwise, `enumerated` does offsets, not indices. It may be better to rename this, or to provide an alternative implementation that actually does use indices and not offsets.

Reply:
Great point; I appreciate the nitty comments like this.